src/ngx_http_lua_bodyfilterby.c

```diff
@@ -272,6 +269,50 @@ ngx_http_lua_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
         return NGX_OK;
     }
 
+    if (in != NULL) {
+        ll = &ctx->in_bufs;
+        while (*ll != NULL) {
+            ll = &(*ll)->next;
+        }
+
+        for (ln = in; ln != NULL; ln = ln->next) {
+            cl = ngx_alloc_chain_link(r->pool);
```

Comment:
We should avoid alloc chain (and free chain below). Small memory blocks can never get freed in nginx memory pools. We should recycle them instead.

Reply:
ngx_alloc_chain_link  allocate chain from the chain pool first and ngx_free_chain free the chain to the chain pool.
ngx_chain_t *
ngx_alloc_chain_link(ngx_pool_t *pool)
{
    ngx_chain_t  *cl;

    cl = pool->chain;

    if (cl) {
        pool->chain = cl->next;
        return cl;
    }

    cl = ngx_palloc(pool, sizeof(ngx_chain_t));
    if (cl == NULL) {
        return NULL;
    }

    return cl;
}

#define ngx_free_chain(pool, cl)                                             \
    (cl)->next = (pool)->chain;                                              \
    (pool)->chain = (cl)